src/NITTA/Model/Networks/Bus.hs

```diff
@@ -755,3 +814,76 @@ instance (UnitTag tag, VarValTime v x t) => Testable (BusNetwork tag v x t) v x
 isDrowAllowSignal Sync = bool2verilog False
 isDrowAllowSignal ASync = bool2verilog True
 isDrowAllowSignal OnBoard = "is_drop_allow"
+-- * Builder
+
+data BuilderSt tag v x t = BuilderSt
+    { signalBusWidth :: Int
+    , availSignals :: [SignalTag]
+    , pus :: M.Map tag (PU v x t)
+    , prototypes :: M.Map tag (PUPrototype tag v x t)
+    }
+
+modifyNetwork net@BusNetwork{bnPus, bnPUPrototypes, bnSignalBusWidth, bnEnv} builder =
+    let st0 =
+            BuilderSt
+                { signalBusWidth = bnSignalBusWidth
+                , availSignals = map (SignalTag . controlSignalLiteral) [bnSignalBusWidth :: Int ..]
+                , pus = bnPus
+                , prototypes = bnPUPrototypes
+                }
+        BuilderSt{signalBusWidth, pus, prototypes} = execState builder st0
+        netIOPorts ps =
+            BusNetworkIO
+                { extInputs = unionsMap puInputPorts ps
+                , extOutputs = unionsMap puOutputPorts ps
+                , extInOuts = unionsMap puInOutPorts ps
+                }
+     in net
+            { bnPus = pus
+            , bnSignalBusWidth = signalBusWidth
+            , bnEnv = bnEnv{ioPorts = Just $ netIOPorts $ M.elems pus}
+            , bnPUPrototypes = prototypes
+            }
+
+defineNetwork bnName ioSync builder = modifyNetwork (busNetwork bnName ioSync) builder
+
+addCustom tag pu ioPorts = do
+    st@BuilderSt{signalBusWidth, availSignals, pus} <- get
+    let ctrlPorts = takePortTags availSignals pu
+        puEnv =
+            def
+                { ctrlPorts = Just ctrlPorts
+                , ioPorts = Just ioPorts
+                , valueIn = Just ("data_bus", "attr_bus")
+                , valueOut = Just (toText tag <> "_data_out", toText tag <> "_attr_out")
+                }
+        pu' = PU pu def puEnv
+        usedPortsLen = length $ usedPortTags ctrlPorts
+    put
+        st
+            { signalBusWidth = signalBusWidth + usedPortsLen
+            , availSignals = drop usedPortsLen availSignals
+            , pus =
+                if M.member tag pus
+                    then error "every PU must has uniq tag"
+                    else M.insert tag pu' pus
```

Comment:
```suggestion
                M.insertWith' (\_ _ -> error "every PU must has uniq tag") tag pu' pus
```

and in same places

Reply:
I've tested this in ghci and got:
`error: Data.Map.insertWith' is gone. Use Data.Map.Strict.insertWith`

At the same time:
```haskell
> import qualified Data.Map.Strict as M
> let c = M.insert 1 "test" M.empty
> let c' = M.insertWith (\_ _ -> error "AAAA") 1 "lol" c -- doesn't throw any error
> c' -- here we will see error "AAAA"
```

So I suggest to leaving this code as it is